File: solutions/c/rational-numbers/1/rational_numbers.c
#include "rational_numbers.h"

#include <stdlib.h>
#include <math.h>

int gcd(int, int);
int gcd(int a, int b) {
    return b == 0 ? abs(a) : gcd(b, a % b);
}

rational_t add(rational_t r1, rational_t r2) {
    rational_t res = {r1.numerator * r2.denominator + r2.numerator * r1.denominator, 
                      r1.denominator * r2.denominator};
    return reduce(res);
}

rational_t subtract(rational_t r1, rational_t r2) {
    rational_t res = {r1.numerator * r2.denominator - r2.numerator * r1.denominator, 
                      r1.denominator * r2.denominator};
    return reduce(res);
}

rational_t multiply(rational_t r1, rational_t r2) {
    rational_t res = {r1.numerator * r2.numerator, 
                      r1.denominator * r2.denominator};
    return reduce(res);
}

rational_t divide(rational_t r1, rational_t r2) {
    rational_t res = {r1.numerator * r2.denominator, 
                      r1.denominator * r2.numerator};
    return reduce(res);
}

rational_t absolute(rational_t r) {
    rational_t res = {abs(r.numerator), abs(r.denominator)};
    return reduce(res);
}

rational_t exp_rational(rational_t r, int16_t n) {
    rational_t res;
    if (n > 0) {
        res = (rational_t) {pow(r.numerator, n), pow(r.denominator, n)};
    } else {
        res = (rational_t) {pow(r.denominator, -n), pow(r.numerator, -n)};
    }
    return reduce(res);
}

float exp_real(float x, rational_t r) {
    return pow(x, (float) r.numerator / r.denominator);
}

rational_t reduce(rational_t r) {
    int divisor = gcd(r.numerator, r.denominator);

    r.numerator /= divisor;
    r.denominator /= divisor;

    if (r.denominator < 0) {
        r.numerator = -r.numerator;
        r.denominator = abs(r.denominator);
    }

    return r;
}

File: solutions/python/rational-numbers/1/rational_numbers.py
from math import gcd


class Rational:
    def __init__(self, numer, denom):
        self.numer = numer
        self.denom = denom
        self.reduce()

    def reduce(self):
        common_divisor = gcd(self.numer, self.denom)
        self.numer //= common_divisor
        self.denom //= common_divisor
        if self.denom < 0:  # Ensure the denominator is positive
            self.numer = -self.numer
            self.denom = -self.denom

    def __eq__(self, other):
        return self.numer == other.numer and self.denom == other.denom

    def __repr__(self):
        return f"{self.numer}/{self.denom}"

    def __add__(self, other):
        return Rational(
            (self.numer * other.denom + other.numer * self.denom),
            (self.denom * other.denom),
        )

    def __sub__(self, other):
        return Rational(
            (self.numer * other.denom - other.numer * self.denom),
            (self.denom * other.denom),
        )

    def __mul__(self, other):
        return Rational((self.numer * other.numer), (self.denom * other.denom))

    def __truediv__(self, other):
        return Rational((self.numer * other.denom), (other.numer * self.denom))

    def __abs__(self):
        return Rational(abs(self.numer), abs(self.denom))

    def __pow__(self, power):
        if power > 0:
            return Rational(pow(self.numer, power), pow(self.denom, power))
        else:
            return Rational(pow(self.denom, -power), pow(self.numer, -power))

    def __rpow__(self, base):
        return pow(base, self.numer) ** (1 / self.denom)
